- [1. 标准输出/标准错误输出/标准输入的默认捕获行为](#1-标准输出标准错误输出标准输入的默认捕获行为)
- [2. 修改和去使能捕获行为](#2-修改和去使能捕获行为)
  - [2.1. 文件描述符级别的捕获行为（默认）](#21-文件描述符级别的捕获行为默认)
  - [2.2. `sys`级别的捕获行为](#22-sys级别的捕获行为)
  - [2.3. 去使能捕获行为](#23-去使能捕获行为)
- [3. 使用`print()`函数调试用例](#3-使用print函数调试用例)
- [4. 在测试用例中访问捕获到的信息](#4-在测试用例中访问捕获到的信息)

## 1. 标准输出/标准错误输出/标准输入的默认捕获行为
在测试执行期间，任何标准输出和标准错误输出都将会被捕获；如果测试失败或者发生异常，异常信息的堆栈也将一同显示，你可以通过`--show-capture`命令行选项来自定义这些行为；

> `--show-capture`的配置项可以为：`no,stdout,stderr,log,all`，默认是`all`；

另外，标准输入被设置为一个`"null"`对象。因为在自动化测试中，很少需要使用到交互输入的场景；

> 实际上，当我们想要使用标准输入时，会得到一个错误：`OSError: reading from stdin while output is captured`；

通常情况下，捕获行为是通过拦截对低级别文件描述符的写入操作来实现的。这就使得我们可以捕获简单的`print()`语句以及测试中子程序的输出行为；


## 2. 修改和去使能捕获行为
`pytest`有两种捕获行为，可以通过`--capture`命令行选项来指定；

### 2.1. 文件描述符级别的捕获行为（默认）
所有向操作系统的文件描述符1（标准输入）和2（标准错误输入）的写入行为都会被捕获，这个也是`pytest`的默认捕获行为，也可以通过`--capture=fd`来指定；

> 文件描述符是与当前进程打开的文件相对应的小整数。例如，标准输入的文件描述符通常是0，标准输出的是1，标准错误的是2，之后被进程打开的文件的描述符依次指定为3、4、5等。 

### 2.2. `sys`级别的捕获行为
只有向`Python`的`sys.stdout`和`sys.stderr`的写入行为会被捕获，不执行对文件描述符的写入的捕获，通过`--capture=sys`来指定；

### 2.3. 去使能捕获行为
通过`--capture=no`可以去使能`pytest`的捕获行为；

也可以通过`-s`命令行选项实现相同的效果，它只是`--capture=no`的一个快捷方式，本质上是一样的；


## 3. 使用`print()`函数调试用例
默认的捕获行为带来的一个主要的好处是，就是可以使用`print()`函数帮助调试用例；

我们来看下面这个例子：

```python
# src/chapter-7/test_module.py

def setup_function(function):
    print("setting up", function)


def test_func1():
    assert True


def test_func2():
    assert False
```

`setup_function(function)`函数会在每个测试用例开始之前执行，做一些初始化的操作；

现在，我们来执行这个模块：

```bash
λ pytest -q src/chapter-7/test_module.py
.F                                                      [100%] 
========================== FAILURES ========================== 
_________________________ test_func2 _________________________

    def test_func2():
>       assert False
E       assert False

src\chapter-7\test_module.py:32: AssertionError
------------------- Captured stdout setup -------------------- 
setting up <function test_func2 at 0x000001F35E76C158>
1 failed, 1 passed in 0.05s
```

可以看到，`pytest`会把失败的用例信息精确的打印出来，并且会忽略其他的用例；


## 4. 在测试用例中访问捕获到的信息
我们可以通过`capsys`、`capsysbinary`、`capfd`和`capfdbinary fixtures`来访问测试执行过程中产生的输出信息；

下面这个例子用于检查测试中的输出信息：

```python
# src/chapter-7/test_output.py

import sys


def test_output(capsys):
    print('hello')
    print('world', file=sys.stderr, end='&')  # 标准错误输出，修改结束符
    captured = capsys.readouterr()
    assert captured.out == 'hello\n'  # print() 默认的结束符是换行符
    assert captured.err == 'world&'
    print('next')
    captured = capsys.readouterr()
    assert captured.out == 'next\n'
```

`readouterr()`方法会返回一个命名元组（包含`out`和`err`属性），表示到目前为止所有的标准输出和标准错误输出，然后**重置缓存区**；

如果你想访问文件描述符级别的测试输出，可以使用`capfd fixture`，它提供了完全相同的接口；

如果想访问的是非文本型的数据，可以使用`capsysbinary fixture`，它的`readouterr()`方法返回的是字节流，参考下面的例子：

```python
# src/chapter-7/test_output.py

def test_binary_output(capsysbinary):
    print('hello')
    captured = capsysbinary.readouterr()
    assert captured.out == b'hello\n'
```

如果你想临时的去使能捕获行为，可以使用`capsys.disabled()`方法，它作为一个上下文管理器来使用，可以禁止`with`作用域中的捕获行为，参考下面的例子：

```python
# src/chapter-7/test_output.py

def test_disabling_capturing(capsys):
    print("hello")
    with capsys.disabled():
        print("world")
    captured = capsys.readouterr()
    assert captured.out == "hello\n"
```
